test/abstract_client.js

```diff
@@ -1418,29 +1438,41 @@ module.exports = function (server, config) {
   describe('unsubscribing', function () {
     it('should send an unsubscribe packet (offline)', function (done) {
       const client = connect()
+      let received = false
 
-      client.unsubscribe('test')
+      client.unsubscribe('test', function (err) {
+        assert.ifError(err)
+        assert(received)
+        client.end(done)
+      })
 
       server.once('client', function (serverClient) {
         serverClient.once('unsubscribe', function (packet) {
           assert.include(packet.unsubscriptions, 'test')
-          client.end(done)
+          received = true
         })
       })
     })
 
     it('should send an unsubscribe packet', function (done) {
       const client = connect()
       const topic = 'topic'
+      let received = false
 
       client.once('connect', function () {
-        client.unsubscribe(topic)
+        client.unsubscribe(topic, function (err) {
+          assert.ifError(err)
+          assert(received)
+          client.end(done)
+        })
       })
 
+      // TODO: all of these test changes have to do with calling end() before the callback is complete. Maybe that should be fixed first
```

Comment:
Would you mind to do this?

Reply:
Depends on what you mean by "this" :)  Do you mean "make these tests better" or do you mean "fix the problems that happen when you call `client.end` from inside a callback?"

IMHO, one of the ugliest problems in this code (MQTT.js), is the spaghetti flow that happens when you call `client.end`.  I've tried to make sense of the order that things get shut down, and there are too many branches and patches and callbacks inside that code to make me confident in _any_ changes to the shutdown code. 

If the flow of `client.end` does get fixed, it needs to be done very carefully. I assume that customers have worked around its oddities (whether they know it or not), and it just really feels like changing how `client.end` works is opening a large can of worms.

Reply:
And I forgot to say that I'm assuming that you want to make these tests better.  Sure, I can do that.

Reply:
> Depends on what you mean by "this" :) Do you mean "make these tests better" or do you mean "fix the problems that happen when you call client.end from inside a callback?"

Sorry for not being clear. I assumed both, anyway I already checked the `end()` code and it's a real mess so we can keep that for another PR 🙏🏼